README.md
# mongoredis
---
一个简洁的基于 Node 的 mongo 及 redis 库（不依赖 mongo 及 redis 原来的应用服务）

## Install
```bash
$ npm install mongoredis
```

### 使用说明
```bash
1. 引用mongoredis
var mgr = require('mongoredis')

2. 创建mongo或redis实例
var redis = mgr.createRedis({
  dbname: 'session',
  dbname: 'user',
  startIndex: 1000000,
  runPeriod: 60 * 1000, // 临时缓存同步时间，默认一分钟，-1 表示关闭
  quickPeriod: -1, // 差异备份时间周期，默认半小时，-1 表示关闭
  fullPeriod: -1, // 全量备份时间周期，默认一天，-1 表示关闭
  fixedPeriod: 30 * 1000, // 数据变动定时检测周期，-1 表示关闭，默认30秒（仅redis有此设置，mongo 无）
  expiredPeriod: 30 * 60 * 1000 // 数据过期检测周期，-1 表示关闭，默认30分（仅redis有此设置，mongo 无）
})

// 注：fixedPeriod参数意义在于 通常redis操作非常频繁，因此采用变更触发的方式会非常耗性能，此时定时检测反倒效率更高；
// 注：定时触发关闭时会自动切换至变更触发状态
// 注：expiredPeriod参数意义在于 通常会将redis用于存储临时数据，因此指定过期时间很有必要，mongo则不然；
// 注：expiredPeriod表示过期定时自动检测的时间周期

var mongo = mgr.createMongo({
  dbname: 'user',
  fullPeriod: 10 * 1000,
  quickPeriod: 3 * 1000
})

3. 操作mongo或redis
redis.set('status', 'logined', 30 * 60 * 1000)
console.log(redis.get('status'))

mongo.insert({
  'username': 'guest',
  'age': 20
}, function(err, result) {
    res.send(result)
})
mongo.getItem(function(row) {
  return String(row.username) === 'guest'
}, function(err, result) {
  res.send(result)
})


```


changelog.txt
### version 1.7.3

Fix JSON.parse('[\uFEFF]') error


### version 1.7.4

Sort suport function


### version 1.7.5

Add upsingle, upsert, upsertsingle method




test.js
/********************************************************************
 * Copyright (C) 2017
 *
 * @author wanghaiyang
 * @date 2017/9/20
 *
 ********************************************************************
 */
 var session = require('./index.js');

console.log('this is ymt session test');

lib/util.js
'use strict';

// [order] 1: asc, -1: desc, 0/'': origin
function sortBy(list, field, order) {
  if (!list || !list.sort || !list.length) return list
  var sort = []
  if (typeof field === 'string') {
    sort.push([field, order === undefined ? 'asc' : String(order)])
  } else if (Object.prototype.toString.call(field) === '[object Array]') {
    sort = field
  }

  for (var i = sort.length - 1; i > -1; i--) {
    if (typeof sort[i] === 'string') {
      sort[i] = [sort[i], 'asc']
    } else if (Object.prototype.toString.call(sort[i]) === '[object Array]') {
      if (typeof sort[i][0] !== 'string') sort[i][0] = String(sort[i][0])
      if (String(sort[i][1]) === '-1' || String(sort[i][1]).toLowerCase() === 'desc') {
        sort[i][1] = 'desc'
      } else if (String(sort[i][1]) === '0') {
        sort[i][1] = '0'
      } else sort[i][1] = 'asc'
    }
  }

  // console.log(sort)
  list.sort(function(a, b) {
    if (!a || !b) return !a ? 1 : -1

    var m, n, result = 0
    for (var i = 0, len = sort.length; i < len; i++) {
      if (sort[i][1] === '0') continue;

      if (Object.prototype.toString.call(a[sort[i][0]]) === '[object Number]' && Object.prototype.toString.call(b[sort[i][0]]) === '[object Number]') {
        m = a[sort[i][0]]
        n = b[sort[i][0]]
      } else {
        m = String(a[sort[i][0]]).toLowerCase()
        n = String(b[sort[i][0]]).toLowerCase()
      }
      result = (m > n ? 1 : m < n ? -1 : 0) * (sort[i][1] === 'desc' ? -1 : 1)

      if (result) break
    }
    return result
  })
  return list
}

function isEqual(a, b, aStack, bStack) {
  // Identical objects are equal. `0 === -0`, but they aren't identical.
  // See the Harmony `egal` proposal: http://wiki.ecmascript.org/doku.php?id=harmony:egal.
  if (a === b) {
    return a !== 0 || 1 / a == 1 / b;
  }
  // A strict comparison is necessary because `null == undefined`.
  if (a === null || b === null || a === undefined || b === undefined) {
    return a === b;
  }
  if (aStack === undefined || bStack === undefined) {
    aStack = [];
    bStack = [];
  }
  // Compare `[[Class]]` names.
  var className = Object.prototype.toString.call(a);
  if (className != Object.prototype.toString.call(b)) {
    return false;
  }
  switch (className) {
    // Strings, numbers, dates, and booleans are compared by value.
    case '[object String]':
      // Primitives and their corresponding object wrappers are equivalent; thus, `"5"` is
      // equivalent to `new String("5")`.
      return a == String(b);
    case '[object Number]':
      // `NaN`s are equivalent, but non-reflexive. An `egal` comparison is performed for
      // other numeric values.
      return a != +a ? b != +b : (a === 0 ? 1 / a === 1 / b : a === +b);
    case '[object Date]':
    case '[object Boolean]':
      // Coerce dates and booleans to numeric primitive values. Dates are compared by their
      // millisecond representations. Note that invalid dates with millisecond representations
      // of `NaN` are not equivalent.
      return +a == +b;
      // RegExps are compared by their source patterns and flags.
    case '[object RegExp]':
      return a.source == b.source &&
        a.global == b.global &&
        a.multiline == b.multiline &&
        a.ignoreCase == b.ignoreCase;
  }
  if (typeof a != 'object' || typeof b != 'object') return false;
  // Assume equality for cyclic structures. The algorithm for detecting cyclic
  // structures is adapted from ES 5.1 section 15.12.3, abstract operation `JO`.
  var length = aStack.length;
  while (length--) {
    // Linear search. Performance is inversely proportional to the number of
    // unique nested structures.
    if (aStack[length] == a) return bStack[length] == b;
  }
  // Add the first object to the stack of traversed objects.
  aStack.push(a);
  bStack.push(b);

  var size = 0,
    result = true;
  // Recursively compare objects and arrays.
  if (className == '[object Array]') {
    // Compare array lengths to determine if a deep comparison is necessary.
    size = a.length;
    result = size == b.length;
    if (result) {
      // Deep compare the contents, ignoring non-numeric properties.
      while (size--) {
        if (!(result = isEqual(a[size], b[size], aStack, bStack))) break;
      }
    }
  } else {
    // Objects with different constructors are not equivalent, but `Object`s
    // from different frames are.
    var aCtor = a.constructor,
      bCtor = b.constructor;
    if (aCtor !== bCtor && !(Object.prototype.toString.call(aCtor) == '[object Function]' && (aCtor instanceof aCtor) &&
        Object.prototype.toString.call(bCtor) == '[object Function]' && (bCtor instanceof bCtor))) {
      return false;
    }
    // Deep compare objects.
    for (var key in a) {
      if (Object.prototype.hasOwnProperty.call(a, key)) {
        // Count the expected number of properties.
        size++;
        // Deep compare each member.
        if (!(result = Object.prototype.hasOwnProperty.call(b, key) && isEqual(a[key], b[key], aStack, bStack))) break;
      }
    }
    // Ensure that both objects contain the same number of properties.
    if (result) {
      for (key in b) {
        if (Object.prototype.hasOwnProperty.call(b, key) && !(size--)) break;
      }
      result = !size;
    }
  }
  // Remove the first object from the stack of traversed objects.
  aStack.pop();
  bStack.pop();

  return result;
}

/**
 * @method hui.formatDate
 * @description 将Date类型解析为String类型. 
 * @param {Date} date 输入的日期
 * @param {String} fmt 输出日期格式
 * @example
 * hui.formatDate(new Date(2006,0,1), 'yyyy-MM-dd HH:mm');
 */
function formatDate(date, fmt) {
  if (!date) date = new Date();
  fmt = fmt || 'yyyy-MM-dd HH:mm:ss';
  var o = {
    'M+': date.getMonth() + 1, //月份      
    'd+': date.getDate(), //日      
    'h+': date.getHours() % 12 === 0 ? 12 : date.getHours() % 12, //小时      
    'H+': date.getHours(), //小时      
    'm+': date.getMinutes(), //分      
    's+': date.getSeconds(), //秒      
    'q+': Math.floor((date.getMonth() + 3) / 3), //季度      
    'S': date.getMilliseconds() //毫秒      
  };
  var week = {
    '0': '/u65e5',
    '1': '/u4e00',
    '2': '/u4e8c',
    '3': '/u4e09',
    '4': '/u56db',
    '5': '/u4e94',
    '6': '/u516d'
  };
  if (/(y+)/.test(fmt)) {
    fmt = fmt.replace(RegExp.$1, (date.getFullYear() + '').substr(4 - RegExp.$1.length));
  }
  if (/(E+)/.test(fmt)) {
    fmt = fmt.replace(RegExp.$1, ((RegExp.$1.length > 1) ? (RegExp.$1.length > 2 ? '/u661f/u671f' : '/u5468') : '') + week[date.getDay() + '']);
  }
  for (var k in o) {
    if (o.hasOwnProperty(k) && new RegExp('(' + k + ')').test(fmt)) {
      fmt = fmt.replace(RegExp.$1, (RegExp.$1.length == 1) ? (o[k]) : (('00' + o[k]).substr(('' + o[k]).length)));
    }
  }
  return fmt;
}
/**
 * @method hui.parseDate
 * @description 将String类型解析为Date类型.  
 * @param {String} fmt 输入的字符串格式的日期
 * @example
 * parseDate('2006-1-1') return new Date(2006,0,1)  
 * parseDate(' 2006-1-1 ') return new Date(2006,0,1)  
 * parseDate('2006-1-1 15:14:16') return new Date(2006,0,1,15,14,16)  
 * parseDate(' 2006-1-1 15:14:16 ') return new Date(2006,0,1,15,14,16);  
 * parseDate('不正确的格式') retrun null  
 */
function parseDate(str) {
  str = String(str).replace(/^[\s\xa0]+|[\s\xa0]+$/ig, '');
  var results = null;

  //秒数 #9744242680 
  results = str.match(/^ *(\d{10}) *$/);
  if (results && results.length > 0)
    return new Date(parseInt(str, 10) * 1000);

  //毫秒数 #9744242682765 
  results = str.match(/^ *(\d{13}) *$/);
  if (results && results.length > 0)
    return new Date(parseInt(str, 10));

  //20110608 
  results = str.match(/^ *(\d{4})(\d{2})(\d{2}) *$/);
  if (results && results.length > 3)
    return new Date(parseInt(results[1], 10), parseInt(results[2], 10) - 1, parseInt(results[3], 10));

  //20110608 1010 
  results = str.match(/^ *(\d{4})(\d{2})(\d{2}) +(\d{2})(\d{2}) *$/);
  if (results && results.length > 5)
    return new Date(parseInt(results[1], 10), parseInt(results[2], 10) - 1, parseInt(results[3], 10), parseInt(results[4], 10), parseInt(results[5], 10));

  //2011-06-08 
  results = str.match(/^ *(\d{4})[\._\-\/\\](\d{1,2})[\._\-\/\\](\d{1,2}) *$/);
  if (results && results.length > 3)
    return new Date(parseInt(results[1], 10), parseInt(results[2], 10) - 1, parseInt(results[3], 10));

  //2011-06-08 10:10 
  results = str.match(/^ *(\d{4})[\._\-\/\\](\d{1,2})[\._\-\/\\](\d{1,2}) +(\d{1,2}):(\d{1,2}) *$/);
  if (results && results.length > 5)
    return new Date(parseInt(results[1], 10), parseInt(results[2], 10) - 1, parseInt(results[3], 10), parseInt(results[4], 10), parseInt(results[5], 10));

  //2011/06\\08 10:10:10 
  results = str.match(/^ *(\d{4})[\._\-\/\\](\d{1,2})[\._\-\/\\](\d{1,2}) +(\d{1,2}):(\d{1,2}):(\d{1,2}) *$/);
  if (results && results.length > 6)
    return new Date(parseInt(results[1], 10), parseInt(results[2], 10) - 1, parseInt(results[3], 10), parseInt(results[4], 10), parseInt(results[5], 10), parseInt(results[6], 10));

  return (new Date(str));
}

module.exports = {
  'sortBy': sortBy,
  'isEqual': isEqual,
  'formatDate': formatDate,
  'parseDate': parseDate
}

lib/mongo.js
'use strict';

var manage = require('./manage')
var Util = require('./util')

module.exports = {
  createNew: function(conf) {
    conf = conf || {}
    if (!conf.dbname || String(conf.dbname).replace(/[0-9a-z_-]+/ig, '')) {
      throw new Error('1st param invalid. [Only 0-9,a-z,_,-]')
    }
    conf.startIndex = conf.startIndex === undefined ? 1000000 : conf.startIndex

    var dbs = manage.createNew({
      dbfile: conf.dbfile || './dbfile/',
      dbname: conf.dbname || 'mongo_' + String(Math.random()).replace('0.', ''),
      startIndex: conf.startIndex,
      runPeriod: conf.runPeriod,
      quickPeriod: conf.quickPeriod,
      fullPeriod: conf.fullPeriod
    })
    dbs.getChangeDiff = function(newdata, olddata) {
      newdata = newdata || []
      olddata = olddata || []

      var keys = {}
      var oldmap = {}
      for (let i = olddata.length - 1; i > -1; i--) {
        oldmap[olddata[i]._id] = olddata[i]
        keys[olddata[i]._id] = true
      }
      var newmap = {}
      for (let i = newdata.length - 1; i > -1; i--) {
        newmap[newdata[i]._id] = newdata[i]
        keys[newdata[i]._id] = true
      }
      var diff = []
      for (let i in keys) {
        if (!Util.isEqual(oldmap[i], newmap[i])) {
          diff.push(JSON.stringify(newmap[i]))
        }
      }

      return diff.join(',\n')
    }


    //  var dataModel = [{
    //  _id: 10001,
    //  id: 100241,
    //  name: 'haiyang',
    //  age: '36'
    // }, {
    //  _id: 10002,
    //  id: 600241,
    //  name: 'qiqi',
    //  age: '9'
    // }]

    // _model 即元数据表， 改了源数据就改了哦
    var dataModel = dbs._model

    var factory = {};
    factory.dbname = conf.dbname;

    factory.getById = function(_id, next) {
      var resp = null
      for (var i = dataModel.length - 1; i > -1; i--) {
        if (String(dataModel[i]._id) === String(_id)) {
          resp = dataModel[i]
          break
        }
      }
      // 深拷贝防止返回结果被误修改
      if (next) next(null, JSON.parse(JSON.stringify(resp)))
    }

    factory.getAll = function(sort, next) {
      if (typeof sort === 'function') {
        next = sort
        sort = null
      }
      var resp = []
      for (var i = dataModel.length - 1; i > -1; i--) {
        resp[i] = dataModel[i]
      }
      if (sort) Util.sortBy(resp, sort)
      if (next) next(null, JSON.parse(JSON.stringify(resp)))
    }

    factory.getItem = function(condition, sort, next) {
      if (typeof sort == 'function') {
        next = sort
        sort = null
      }
      var resp = []
      if (typeof condition === 'function') {
        var list = sort ? Util.sortBy(dataModel, sort) : dataModel
        for (var i = 0, len = list.length; i < len; i++) {
          if (condition(list[i])) {
            resp.push(list[i])
            break
          }
        }
      }
      resp = resp[0] || null
      if (next) next(null, JSON.parse(JSON.stringify(resp)))
    }

    // page: 1, 2, 3, ..., n or condition
    // factory.getItems = function(condition, sort, page, count, next) {

    factory.getItems = function(condition, sort, page, count, next) {
      if (typeof sort === 'function') {
        next = sort
        sort = null
      }
      var resp = []
      if (typeof condition === 'function') {
        var list = sort ? Util.sortBy(dataModel, sort) : dataModel
        for (var i = 0, len = list.length; i < len; i++) {
          if (condition(list[i])) {
            resp.push(list[i])
          }
        }
      }
      if (sort && typeof(sort[0]) == 'function') {
        resp.sort(sort[0])
      } else if (sort) Util.sortBy(resp, sort)

      var pp = 1
      if (typeof page == 'function') {
        pp = page(resp) || 1
      } else {
        pp = parseInt(page) == page ? page : 0
        pp = pp < 1 ? 1 : pp > 2147483646 ? 2147483646 : pp
      }

      count = parseInt(count) == count ? count : 0
      count = count < 1 ? 1 : count > 2147483647 ? 2147483647 : count

      if (next) next(null, JSON.parse(JSON.stringify(resp.slice((pp - 1) * count, pp * count))), resp.length, pp)
    }

    factory.count = function(condition, next) {
      var num = -1
      if (typeof condition === 'function') {
        num = 0
        for (var i = dataModel.length - 1; i > -1; i--) {
          if (condition(dataModel[i])) {
            num++
          }
        }
      }
      if (next) next(null, num)
    }

    factory.insert = function(item, update, opt, next) {
      if (typeof opt === 'function') next = opt
      if (!next) {
        next = update
        update = null
      }
      if (!item) {
        if (next) next('Insert "Item" is null', null);
      }
      var row = {}
      for (var i in item) {
        if (!item.hasOwnProperty(i)) continue;
        row[i] = item[i]
      }
      // 注：处理自增属性，先赋一个默认值startIndex
      var autoIndent = { _id: conf.startIndex }
      if (opt && opt.auto_increment) opt.auto_increment.forEach(function(item1) {
        autoIndent[item1] = conf.startIndex
      })
      // 注：遍历整张表如果发现比默认值还大则在最大值基础上再加一
      for (var key in autoIndent) {
        if (!autoIndent.hasOwnProperty(key)) continue;
        for (i = dataModel.length - 1; i > -1; i--) {
          if (parseInt(dataModel[i][key], 10) > autoIndent[key]) {
            autoIndent[key] = parseInt(dataModel[i][key], 10)
          }
        }
        row[key] = String(autoIndent[key] + 1)
      }

      if (typeof update === 'function') {
        update(row)
      }

      dataModel.push(row)
      var resp = row

      dbs.checkModify()
      if (next) next(null, JSON.parse(JSON.stringify(resp)))
    }

    // http://docs.factorydb.org/manual/reference/method/db.collection.update/#update-parameter
    // {
    //    upsert: <boolean>,
    //    multi: <boolean>
    //  }
    factory.update = function(condition, update, opt, next) { // multi 默认 true, upsert 默认 false
      var err = null
      if (typeof condition !== 'function') err = '1st param should be Fuction'
      if (typeof update !== 'function') err = '2nd param should be Fuction'
      if (typeof opt === 'function') {
        next = opt
        opt = null
      }

      var resp = null
      var exist = false
      if (typeof condition === 'function' && typeof update === 'function') {
        resp = []
        var list = opt && opt.sort ? Util.sortBy(dataModel, opt.sort) : dataModel
        for (var i = 0, len = list.length; i < len; i++) {
          if (condition(list[i])) {
            var _id = list[i]._id
            update(list[i])
            list[i]._id = _id
            resp.push(list[i])
            exist = true
            if (opt && opt.multi === false) break
          }
        }
        // Todo: test it
        if (opt.upsert && !exist) {
          var row = {}
          factory.insert(row, function(item) {
            update(item)
          }, opt, function(err, item) {
            resp.push(item)
          })
        }
      }

      dbs.checkModify()
      if (next) next(err, JSON.parse(JSON.stringify(resp)))
    }

    factory.upsingle = function(condition, update, opt, next) { // multi: false
      if (typeof opt === 'function') {
        next = opt
        opt = null
      }
      opt = opt || {}
      opt.multi = false
      return factory.update(condition, update, opt, function(err, doc) {
        next(err, doc && doc[0] ? doc[0] : null)
      })
    }

    factory.upsert = function(condition, update, opt, next) { // upsert: false
      if (typeof opt === 'function') {
        next = opt
        opt = null
      }
      opt = opt || {}
      opt.upsert = true
      return factory.update(condition, update, opt, next)
    }

    factory.upsertsingle = function(condition, update, opt, next) { // upsert: true, multi: false
      if (typeof opt === 'function') {
        next = opt
        opt = null
      }
      opt = opt || {}
      opt.upsert = true
      opt.multi = false
      return factory.update(condition, update, opt, function(err, doc) {
        next(err, doc && doc[0] ? doc[0] : null)
      })
    }

    factory.updateById = function(_id, update, next) {
      var resp = null
      var err = null
      if (typeof update === 'function') {
        resp = []
        for (var i = dataModel.length - 1; i > -1; i--) {
          if (String(dataModel[i]._id) === String(_id)) {
            update(dataModel[i])
            resp.push(dataModel[i])
            break
          }
        }
      } else err = '2nd param is not Function'

      dbs.checkModify()
      if (next) next(err, JSON.parse(JSON.stringify(resp)));
    }

    factory.remove = function(condition, next) {
      var resp = []
      if (typeof condition === 'function') {
        for (var i = 0, len = dataModel.length; i < len; i++) {
          if (condition(dataModel[i])) {
            dataModel[i] = {
              '_id': dataModel[i]._id
            }
            resp.push(dataModel[i])
          }
        }
      }

      dbs.checkModify()
      if (next) next(null, JSON.parse(JSON.stringify(resp)))
    }

    return factory
  }
}

lib/manage.js
/* global module */
'use strict';
var fs = require('fs')
var path = require('path')
var Util = require('./util')

var JSONParse = function(str) {
  var result
  try {
    result = JSON.parse(str)
  } catch (e) {
    result = Function('return ' + str)()
  }
  return result
}

module.exports = {
  createNew: function(conf) {
    conf = conf || {}
    if (!conf.dbname || String(conf.dbname).replace(/[0-9a-z_-]+/ig, '')) {
      throw new Error('1st param invalid. [Only 0-9,a-z,_,-]')
    }
    var mdb = {
      config: {
        path: conf.dbfile || './dbfile/',
        dbname: conf.dbname || 'tmp',
        startIndex: conf.startIndex === undefined ? 1000000 : conf.startIndex,
        runPeriod: conf.runPeriod || 60 * 1000,
        quickPeriod: conf.quickPeriod || 30 * 60 * 1000,
        fullPeriod: conf.fullPeriod || 24 * 60 * 60 * 1000,
        curbak: 'a'
      },
      _origin: [],
      _model: [
        // {_id: 10001, key: 'tom', value: '10000'},
        // {_id: 10002, key: 'dim', value: '10001'}
      ],
      runTimer: null,
      fullTimer: null,
      quickTimer: null
    }

    // public
    mdb.checkModify = function() {
      if ([conf.runPeriod, conf.quickPeriod, conf.fullPeriod].join('') === '-1-1-1') return ''
      // 注：1. 配合差量备份 2. 防止启动后无任何数据变更也频繁存储
      if (!mdb.getChangeDiff(mdb._model, mdb._origin)) return ''

      // 检查运行备份任务是否在运行
      if (!mdb.runTimer && String(conf.runPeriod) !== '-1') {
        mdb.runTimer = setTimeout(function(mdb) {
          mdb.runTimer = null
          mdb.saveRunFile()
        }.bind({}, mdb), mdb.config.runPeriod)
      }
      // 检查差量备份任务是否在运行
      if (!mdb.quickTimer && String(conf.quickPeriod) !== '-1') {
        mdb.quickTimer = setTimeout(function(mdb) {
          mdb.quickTimer = null
          mdb.saveQuickFile()
        }.bind({}, mdb), mdb.config.quickPeriod)
      }
      // 检查全量备份任务是否在运行
      if (!mdb.fullTimer && String(conf.fullPeriod) !== '-1') {
        mdb.fullTimer = setTimeout(function(mdb) {
          mdb.fullTimer = null
          mdb.saveFullFile()
        }.bind({}, mdb), mdb.config.fullPeriod)
      }
    }

    // public
    mdb.getChangeDiff = function(newdata, olddata) {
      newdata = newdata || []
      olddata = olddata || []

      var diff = []
      for (let i = 0, len = newdata.length; i < len; i++) {
        if (JSON.stringify(newdata[i]) !== JSON.stringify(olddata[i])) {
          diff.push(JSON.stringify(newdata[i]))
        }
      }

      return diff.join(',\n')
    }
    // private
    mdb.getFullContent = function() {
      var result = []
      var list = Util.sortBy(mdb._model, '_id', 'asc')
      for (let i = 0, len = list.length; i < len; i++) {
        result.push(JSON.stringify(list[i]))
      }
      return result.join(',\n')
    }

    // private
    mdb.saveRunFile = function() {
      mdb.readRunFile(function(res) {
        var str = mdb.getFullContent()
        // 有变动才存盘！否则在定时任务下会产生大量重复冗余文件
        if (!Util.isEqual(JSONParse('[' + str + ']'), JSONParse('[' + res + ']'))) {
          var curbak = mdb.config.curbak
          // update db file
          var fpath1 = path.normalize(mdb.config.path + mdb.config.dbname + '_' + curbak + '_run.txt')
          fs.writeFile(fpath1, str, 'utf8', function(err) {
            if (err) return console.error(err)
            // 这里无需切换 mdb.config.curbak，前面的 mdb.readRunFile 中会自动切换 curbak
            console.log(mdb.config.dbname + ' => Full(tmp) save success 1. ' + fpath1)
            var fpath2 = path.normalize(mdb.config.path + mdb.config.dbname + '_' + (curbak == 'a' ? 'b' : 'a') + '_run.txt')
            fs.writeFile(fpath2, str, 'utf8', function(err) {
              if (err) return console.error(err)
              // 这里无需切换 mdb.config.curbak，前面的 mdb.readRunFile 中会自动切换 curbak
              console.log(mdb.config.dbname + ' => Full(tmp) save success 2. ' + fpath2)
            })
          })
        }
      })
    }
    // private
    mdb.saveQuickFile = function() {
      var result = mdb.getChangeDiff(mdb._model, mdb._origin)
      if (result) {
        // 重置_origin[重要!!]
        mdb._origin = JSONParse(JSON.stringify(mdb._model))

        var fpath = path.normalize(mdb.config.path + 'diff_' + mdb.config.dbname + '_' + Util.formatDate(new Date(), 'yyyyMMddHHmmss') + '.txt')
        fs.writeFile(fpath, result, 'utf8', function(err) {
          if (err) return console.error(err)
          console.log(mdb.config.dbname + ' => Quick save success. ' + fpath)
        })
      }
    }
    // private
    mdb.saveFullFile = function() {
      if (mdb.quickTimer) {
        clearTimeout(mdb.quickTimer)
        mdb.quickTimer = null
        mdb.saveQuickFile()
      }

      mdb.readRunFile(function(res) {
        var str = mdb.getFullContent()
        // 有变动才存盘！否则在定时任务下会产生大量重复冗余文件
        if (!Util.isEqual(JSONParse('[' + str + ']'), JSONParse('[' + res + ']'))) {
          var fpath = path.normalize(mdb.config.path + 'full_' + mdb.config.dbname + '_' + Util.formatDate(new Date(), 'yyyyMMddHHmmss') + '.txt')
          fs.writeFile(fpath, str, 'utf8', function(err) {
            if (err) return console.error(err)
            mdb._origin = JSONParse('[' + str + ']')
            console.log(mdb.config.dbname + ' => Full(timestamp) save success. ' + fpath)
          })
        }
      })
    }
    // private
    mdb.readRunFile = function(callback) {
      // 注：使用两个文件代替原来单个文件的方式，主要是防止存储文件时程序异常退出，导致数据文件变成空文件丢失所有数据的情况
      // 注：不过使用双文件备份后，仍会丢失上次到本次保存之间的差集，这点需要特别注意
      var mt1 = false
      var mt2 = false
      var handler = function() {
        if (mt1 !== false && mt2 !== false) {
          mdb.config.curbak = mt1 > mt2 ? 'b' : 'a'
          if (!mt1 && !mt2) callback('')
          else fs.readFile(mt1 > mt2 ? url1 : url2, 'utf8', function(err, fileContent) {
            if (err) return console.error(err)
            callback(fileContent)
          })
        }
      }
      // return [{key: 1, value: 11},{key: 2, value: 22},{key: 3, value: 33}]
      var url1 = path.normalize(mdb.config.path + mdb.config.dbname + '_a_run.txt')
      fs.exists(url1, function(exists) {
        if (exists) {
          fs.stat(url1, function(err, states) {
            if (err) return console.error(err)
            mt1 = states.mtime.getTime()
            handler()
          })
        } else {
          mt1 = ''
          handler()
        }
      })
      var url2 = path.normalize(mdb.config.path + mdb.config.dbname + '_b_run.txt')
      fs.exists(url2, function(exists) {
        if (exists) {
          fs.stat(url2, function(err, states) {
            if (err) return console.error(err)
            mt2 = states.mtime.getTime()
            handler()
          })
        } else {
          mt2 = ''
          handler()
        }
      })
    }

    // private
    mdb.init = function() {
      mdb.readRunFile(function(res) {
        var olddata = JSONParse('[' + res + ']')
        for (let i = olddata.length - 1; i > -1; i--) {
          var str = JSON.stringify(olddata[i])
          mdb._model[i] = JSONParse(str)
          mdb._origin[i] = JSONParse(str)
        }
      })
    }
    mdb.init()

    return mdb
  }
}

lib/redis.js
'use strict';

var manage = require('./manage')
var Util = require('./util')

module.exports = {
  createNew: function(conf) {
    conf = conf || {}
    if (!conf.dbname || String(conf.dbname).replace(/[0-9a-z_-]+/ig, '')) {
      throw new Error('1st param invalid. [Only 0-9,a-z,_,-]')
    }
    conf.startIndex = conf.startIndex === undefined ? 1000000 : conf.startIndex

    var dbs = manage.createNew({
      dbfile: conf.dbfile || './dbfile/',
      dbname: conf.dbname || 'redis_' + String(Math.random()).replace('0.', ''),
      startIndex: conf.startIndex,
      runPeriod: conf.runPeriod,
      quickPeriod: conf.quickPeriod,
      fullPeriod: conf.fullPeriod
    })
    dbs.getChangeDiff = function(newdata, olddata) {
      newdata = newdata || []
      olddata = olddata || []

      var keys = {}
      var newmap = {}
      for (let i = newdata.length - 1; i > -1; i--) {
        newmap[newdata[i].key] = newdata[i]
        keys[newdata[i].key] = true
      }
      var oldmap = {}
      for (let i = olddata.length - 1; i > -1; i--) {
        oldmap[olddata[i].key] = olddata[i]
      }
      var diff = []
      for (let i in keys) {
        if (!oldmap[i] || oldmap[i].expired != newmap[i].expired || !Util.isEqual(oldmap[i].value, newmap[i].value)) {
          diff.push(JSON.stringify(newmap[i]))
        }
      }

      return diff.join(',\n')
    }

    var monitor = {
      expiredPeriod: (conf.expiredPeriod || 30 * 60 * 1000),
      expiredTimer: null
    }

    //  var dataModel = [{
    //  _id: 10001,
    //  id: 100241,
    //  key: 'haiyang',
    //  value: '36'
    // }, {
    //  _id: 10002,
    //  id: 600241,
    //  key: 'qiqi',
    //  value: '9'
    // }]
    var dataModel = dbs._model

    var factory = {};
    factory.dbname = conf.dbname;

    factory.get = function(key) {
      key = String(key)
      var value = ''
      var now = (new Date()).getTime()
      for (let i = dataModel.length - 1; i > -1; i--) {
        if (dataModel[i].key === key) {
          if (!dataModel[i].expired || Util.parseDate(dataModel[i].expired).getTime() > now) {
            value = dataModel[i].value
          }
          break
        }
      }

      if (conf.expiredPeriod) monitor.checkExpired()
      if (!conf.fixedPeriod) dbs.checkModify()
      return value
    }

    factory.set = function(key, value, expired) {
      if (value === null || value === undefined) value = ''
      // 注释掉原因：value相同，expired不同，也需要更新数据文件
      // if (Util.isEqual(factory.get(key), value)) return;

      key = String(key)
        // value = JSON.stringify(value)

      expired = expired ? Util.formatDate(Util.parseDate((new Date()).getTime() + expired)) : ''

      var _id = conf.startIndex
      for (var i = dataModel.length - 1; i > -1; i--) {
        if (parseInt('0' + String(dataModel[i]._id), 10) > _id) {
          _id = parseInt('0' + String(dataModel[i]._id), 10)
        }
        if (dataModel[i].key === key) {
          dataModel[i].value = value
          dataModel[i].expired = expired
          break
        }
      }
      if (i < 0) { // not exist
        dataModel.push({
          '_id': String(_id + 1),
          'key': key,
          'value': value,
          'expired': expired
        })
      }

      if (conf.expiredPeriod) monitor.checkExpired()
      if (!conf.fixedPeriod) dbs.checkModify()
    }

    factory.clear = function(condition) {
      for (var i = dataModel.length - 1; i > -1; i--) {
        if (typeof condition === 'function' && !condition(dataModel[i])) continue
        dataModel[i] = {
          '_id': dataModel[i]._id
        }
      }

      if (conf.expiredPeriod) monitor.checkExpired()
      if (!conf.fixedPeriod) dbs.checkModify()
    }

    monitor.checkExpired = function() {
      if (monitor.expiredTimer) return;
      monitor.expiredTimer = setTimeout(function(monitor, dataModel) {
        monitor.expiredTimer = null

        var now = (new Date()).getTime()
        for (let i = dataModel.length - 1; i > -1; i--) {
          if (dataModel[i].expired && Util.parseDate(dataModel[i].expired).getTime() < now) {
            dataModel[i] = {
              '_id': dataModel[i]._id
            }
          }
        }
      }.bind({}, monitor, dataModel), monitor.expiredPeriod)
    }

    if (conf.fixedPeriod && !String(conf.fixedPeriod).replace(/\d+/, '') && Number(conf.fixedPeriod)) {
      conf.fixedPeriod = Number(conf.fixedPeriod)
      
      factory.fixedTimer = setInterval(function(dbs) {
        dbs.checkModify()
      }.bind({}, dbs), conf.fixedPeriod)
    } else {
      conf.fixedPeriod = ''
    }

    if (conf.expiredPeriod && !String(conf.expiredPeriod).replace(/\d+/, '') && Number(conf.expiredPeriod)) {
      conf.expiredPeriod = Number(conf.expiredPeriod)
    } else {
      conf.expiredPeriod = ''
    }

    return factory
  }
}